src/modules/scopes.js

```diff
@@ -72,63 +72,74 @@ function decorateRootScope($delegate, $parse) {
 
-  // var _watch = scopePrototype.$watch;
-  // scopePrototype.$watch = function (watchExpression, reactionFunction) {
-  //   var watchStr = humanReadableWatchExpression(watchExpression);
-  //   var scopeId = this.$id;
-  //   if (typeof watchExpression === 'function') {
-  //     arguments[0] = function () {
-  //       var start = perf.now();
-  //       var ret = watchExpression.apply(this, arguments);
-  //       var end = perf.now();
-  //       hint.emit('scope:watch', {
-  //         id: scopeId,
-  //         watch: watchStr,
-  //         time: end - start
-  //       });
-  //       return ret;
-  //     };
-  //   } else {
-  //     var thatScope = this;
-  //     arguments[0] = function () {
-  //       var start = perf.now();
-  //       var ret = thatScope.$eval(watchExpression);
-  //       var end = perf.now();
-  //       hint.emit('scope:watch', {
-  //         id: scopeId,
-  //         watch: watchStr,
-  //         time: end - start
-  //       });
-  //       return ret;
-  //     };
-  //   }
-
-  //   if (typeof reactionFunction === 'function') {
-  //     var applyStr = reactionFunction.toString();
-  //     arguments[1] = function () {
-  //       var start = perf.now();
-  //       var ret = reactionFunction.apply(this, arguments);
-  //       var end = perf.now();
-  //       hint.emit('scope:reaction', {
-  //         id: this.$id,
-  //         watch: watchStr,
-  //         time: end - start
-  //       });
-  //       return ret;
-  //     };
-  //   }
-
-  //   return _watch.apply(this, arguments);
-  // };
+  var _watch = scopePrototype.$watch;
+  var _digestEvents = [];
+  scopePrototype.$watch = function (watchExpression, reactionFunction) {
+    var watchStr = humanReadableWatchExpression(watchExpression);
+    var scopeId = this.$id;
+    if (typeof watchExpression === 'function') {
+      arguments[0] = function () {
+        var start = perf.now();
+        var ret = watchExpression.apply(this, arguments);
+        var end = perf.now();
+        _digestEvents.push({
+          eventType: 'scope:watch',
+          id: scopeId,
+          watch: watchStr,
+          time: end - start
+        });
+        return ret;
+      };
+    } else {
+      var thatScope = this;
+      arguments[0] = function () {
+        var start = perf.now();
+        var ret = thatScope.$eval(watchExpression);
+        var end = perf.now();
+        _digestEvents.push({
+          eventType: 'scope:watch',
+          id: scopeId,
+          watch: watchStr,
+          time: end - start
+        });
+        return ret;
+      };
+    }
+
+    if (typeof reactionFunction === 'function') {
+      arguments[1] = function () {
+        var start = perf.now();
+        var ret = reactionFunction.apply(this, arguments);
+        var end = perf.now();
+        _digestEvents.push({
+          eventType: 'scope:reaction',
+          id: this.$id,
+          watch: watchStr,
+          time: end - start
+        });
+        return ret;
```

Comment:
i see. something like this(below) wouldn't work? (or are there perf implications, or would it affect the timer maybe?)  

    function someHelper(eventType, id, watStr, getRet) {
      return function() {
        var start = perf.now();
        var ret = getRet();
        var end = perf.now();
        _digestEvents.push({
          eventType: 'scope:reaction',
          id: this.$id,
          watch: watchStr,
          time: end - start
        });
      };
    }

    arguments[0] = someHelper('scope:watch', id, watchStr, (function(args) {
      return watchExpression.apply(this, args);
    }).bind(this, arguments));

    arguments[0] = someHelper('scope:watch', scopeId, watchStr, (function(watchExpression) {
      return this.$eval(watchExpression);
    }).bind(thatScope, watchExpression));

    arguments[1] = someHelper('scope:reaction', this.$id, watchStr, (function(args) {
      return reactionFunction.apply(this, args);
    }).bind(this, arguments));